spring-integration-core/src/main/java/org/springframework/integration/util/MessagingMethodInvokerHelper.java

```diff
@@ -277,13 +292,15 @@ private void prepareEvaluationContext() {
 			}
 		}
 		else {
-			AnnotatedMethodFilter filter = new AnnotatedMethodFilter(this.annotationType, this.methodName,
-					this.requiresReply);
-			Assert.state(canReturnExpectedType(filter, targetType, context.getTypeConverter()),
+			AnnotatedMethodFilter annotatedMethodFilter = new AnnotatedMethodFilter(this.annotationType,
+					this.methodName, this.requiresReply);
+			Assert.state(canReturnExpectedType(annotatedMethodFilter, targetType, context.getTypeConverter()),
 					"Cannot convert to expected type (" + this.expectedType + ") from " + this.method);
-			context.registerMethodFilter(targetType, filter);
+			context.registerMethodFilter(targetType, annotatedMethodFilter);
+			declaredMethodResolver.registerMethodFilter(targetType, annotatedMethodFilter);
 		}
 		context.setVariable("target", this.targetObject);
+		context.setMethodResolvers(Collections.singletonList(declaredMethodResolver));
```

Comment:
OK. I can accept your suggestion as a change, but what you say if you know that `evaluationContext` is `prototype`. So, we don't effect any other component with this replacement.